Construct graph from a few more input data structures (mimicking networkx)

iosonofabio opened the pull request:
Aiming to fix #433 and related forum request, and matching (roughly) networkx's range of input fuctions:

- `Graph.ListDict`
- `Graph.DictDict`

(we already had:

- `Graph.TupleList`
- `Graph.DictList`

)

Notes:
- currently not working for lazy iterators rather than dicts. Would be easy (if a little obscuring) to extend, if you think it'd be useful (can't think of a clear case)
- poorly documented ATM, but has some tests
- `Graph.ListDict` currently takes no edge attributes, `Graph.DictDict` is better for that purpose

@szhorvat @ntamas do you feel this addresses the issue? If so I'd be happy to improve docs.

TODO:
 - [x] finish docstrings
 - [x] add API docs/user guide
 - [x] extend tests to cover more cases
 - [x] extend functionality (edge attributes,etc)
 - [x] decide what to do with `"name"` attribute: fixed or not (API break)?

vtraag commented on src/igraph/__init__.py:
```diff
@@ -3205,6 +3205,71 @@ def TupleList(
+            name_map = {}
+            edge_list = []
+            for item, sequence in edges.items():
+                vid = name_map[item] = name_map.get(item, n+1)
```
There's also an internal type that generates unique (integer) identifiers: [`UniqueIdGenerator`](https://igraph.org/python/doc/api/igraph.datatypes.UniqueIdGenerator.html). That should simplify it to:

```python
for item, sequence in edges.items():
    for item2 in sequence:
        edge_list.append(id[item], id[item2])
```

iosonofabio replied:
good point

vtraag commented on src/igraph/__init__.py:
```diff
@@ -3205,6 +3205,71 @@ def TupleList(
+        cls,
+        edges,
+        directed=False,
+        vertex_name_attr="name",
```
I think the vertex name should always be called `name`, otherwise `igraph` would not recognize it as a proper "name" (i.e. including the indexing on the name), right?

ntamas replied:
No, I think it's okay this way; there could be perfectly valid reasons for trying to construct a graph where the `name` attribute would be provided later by the user and the original names are stored in a different attribute (e.g., `original_name`).

vtraag reviewed (commented):
Great! Looks good to me, I have a small suggestion for using an internal type that slightly simplifies the code, and I'm not sure if we should make the name attribute a choice, but perhaps @ntamas has some comments about it.

We could take a look at all other [input functions](https://networkx.org/documentation/stable/reference/convert.html) that `networkx` supports, it would be nice to also support the other options.

iosonofabio commented:
Ok, I moved a bunch of constructor functions into a separate module called `graph_constructors.py` and added a variation to `Adjacency` that pairs up with networkx's ability to read pandas DataFrame adjacency matrices.

I think we are now covering all the input options they cover.

Docs on arguments for some of the new constructors are still poor, happy to improve in the next few commits.

iosonofabio commented:
Since I was there I also moved a bunch of constructors (almost all, in fact) into the new module, which reduces the size of `__init__.py` by ~1,000 lines.

Ideally we would actually have separate modules for say constructing from file versus from live objects, but that seems a little much for this humble PR, so let's plan that for after this one is merged unless you guys have any urgency, in which case we can do it here as well (I can, or you can push here @ntamas).

ntamas commented:
> Since I was there I also moved a bunch of constructors (almost all, in fact) into the new module, which reduces the size of `__init__.py` by ~1,000 lines.

Yay, that's definitely an improvement, thanks!

Ideally, in the long term (i.e. the duration of the CZI grant) we should be moving towards having many small modules grouped around topics instead of a single monolithic `Graph` class that tries to cover everything with instance and class methods. Of course that completely breaks the API of the module and I'm not sure yet how to do this nicely while also providing some kind of a backwards compatibility mechanism to ease the transition.

Maybe now that we also have the `igraph` module on PyPI, we could use _that_ for a new-style Python API that is better designed than what we have now, and keep `python-igraph` with the old monolithic Graph class instead? That would also save the hassle of trying to keep `igraph` and `python-igraph` in sync while people get used to using `pip install igraph` instead of `pip install python-igraph`; basically old projects can stick to `python-igraph`, new projects can use `igraph` (with a new, improved API), and old projects who eventually bite the bullet and migrate to the new API can make the switch whenever it's convenient for them.

iosonofabio commented:
Love that idea! I suggest a two step process:

1. We split the implementation of all/almost all graph methods into modules, but still link them in so the API didn't change.

2. We fork that repo and bye bye

ntamas commented:
> We split the implementation of all/almost all graph methods into modules, but still link them in so the API didn't change.

Yes, I think this is doable; I think if you implement a method like

```
def do_something(graph, ...):
    ...
```

then you can link it to the Graph class with `Graph.do_something = do_something` and then it also becomes an instance method of the `Graph` class; so, we can start stripping methods separately into functions and link them back to the `Graph` class to keep compaibility.

Still, there are too many parallel things going on in the `python-igraph` repo at the moment (the integer transition, the plotting backends, plus this one) so I'm going to put this train of thought aside for the moment until we finish the other two (which mostly involves me finally finishing the int conversion by tidying up the types in the C glue code, and  then merging the plotting PR).

iosonofabio commented:
Don't worry, I'll just chug along against develop now that we have it, and this PR can come after those are done

ntamas commented on setup.py:
```diff
@@ -821,7 +821,13 @@ def use_educated_guess(self):
+        "igraph.drawing",
+        "igraph.remote",
+        "igraph.constructors",
+        ],
```
Nowadays I tend to delegate all formatting-related decisions to [`black`](https://black.readthedocs.io/en/stable/) so I could free up some mental capacity to focus on more important things. Eventually, all of `python-igraph`'s source code will be piped through `black` and there will be a pre-commit hook that runs `black` on changed files before committing. I'm just mentioning this so you won't put too much effort into formatting as `black` will take care of it anyway. 

In this particular case, `black` will dedent the closing square bracket so it lines up with `packages=` above.

iosonofabio replied:
you're right, sorry about that, just habit...

ntamas replied:
No problem, just wanted to mention that you shouldn't spend too much time on code formatting :)

ntamas commented on src/igraph/constructors/from_object.py:
```diff
@@ -0,0 +1,462 @@
+    # Check if we are using names or integers
+    for item in edges:
+        break
+    if not isinstance(item, (int, str)):
```
This will blow up if `edges` is an empty iterable; it should be handled separately.

I have found the following helper function useful in such cases:

```
def first(iterable, default=None):
    for item in iterable:
        return item
    return default
```

and then you can simply write:

```
if not isinstance(first(edges, default=0), (int, str)):
    raise ValueError("Keys must be integers or strings")
```

and this will handle the case of an empty iterable correctly.

iosonofabio replied:
Good catch, I fixed the corner case now

ntamas commented on src/igraph/constructors/from_object.py:
```diff
@@ -0,0 +1,462 @@
+    - Alice - David (with weight 2)
+    """
+    # Check if we are using names or integers
+    for item in edges:
```
Same here; this won't work if `edges` is empty because `item` will be unassigned.

iosonofabio replied:
yep, fixed now

ntamas commented on src/igraph/constructors/from_file.py:
```diff
@@ -0,0 +1,228 @@
+    """Constructs a graph based on an adjacency matrix from the given file.
+
+    Additional positional and keyword arguments not mentioned here are
+    passed intact to L{Adjacency}.
```
`L{Adjacency}` is a broken reference now that we stripped this function out of the `Graph` class.

ntamas commented:
Haven't reviewed the whole PR yet but I've added some comments that are worth considering.

One general problem I have with the new `construct_...` methods is that it is unclear whether they are meant to be part of the public API. If they are, then having `cls` as the first argument makes the signature quite awkward; I'd rather have it as a keyword-only argument (now that Python finally has keyword-only arguments) like this:

```python
def construct_graph_somehow(first_arg, second_arg, *, graph_factory=Graph):
    ...
```

and then simulate the "old" class methods of the Graph class with another decorator:

```python
from functools import wraps

def igraph_classmethod(func):
    @wraps(func)
    def wrapped(cls, *args, **kwds):
        kwds["graph_factory"] = cls
        return func(*args, **kwds)
    return wrapped
```

The problem with this proposal is that we will very likely introduce circular imports by having `Graph` as the default argument for `graph_factory`. A workaround is to let the default value of `graph_factory` be `None` and use a deferred import inside the constructor function to replace `None` with `Graph` if needed.

ntamas commented:
Also, when reviewing the PR, I will simply assume that you did not modify the bodies of the functions that you moved out from the `Graph` class and only added a few additional functions on your own so I won't need to compare the internals of the functions that were moved out with their original versions. Is that correct?

iosonofabio commented:
Thanks @ntamas 

1. yes I didn't touch the body of other functions.
2. Sorry about the linting, sometimes it's too strong a temptation to resist ;-) Yes let's just black-lint everything at the end
3. I wouldn't add those functions to the API: there should be **one** preferred way to create graphs and that's currently via class methods of `Graph`. In fact, I would like them to intentionally NOT be part of the API for that reason (perhaps I could prefix the function names with `_` to make sure people don't stumble upon them.

I checked if `help(Graph.Adjacency)` and other class/instance methods yield the right documentation now, and they do. So I'm not sure `L{Adjacency}` is a broken reference?

ntamas commented:
> I wouldn't add those functions to the API: there should be one preferred way to create graphs and that's currently via class methods of Graph. In fact, I would like them to intentionally NOT be part of the API

Fair enough, you have a point here; in this case, let's mark these methods with a leading underscore so it is clear that they are internal. I'm afraid that PyDoctor would pick them up and include them in the documentation if we don't do so, and maybe people coming from NetworkX would feel a temptation to import these functions directly instead of relying on the class methods.

iosonofabio commented:
I should add export functions for dict of dicts and friends

iosonofabio commented:
Ok now we have export functions for each import function in IO, which should cover everything networkx does and then some. This PR is somewhat sensitive since it generates a decent amonut of new API estate, so it'd be great if @vtraag you could have a look

If you guys think it's ok I can add proper docs and then we can merge.

iosonofabio commented:
I included PR #408 in this one since they would conflict and it's a closely related task. Closing the other one.

iosonofabio commented:
Extended tests a little including some trivial cases, added docs and docstrings, and some flexibility as of the "name" attribute, should be ready for merge

vtraag reviewed (commented):
Great to see all these input/output structure being supported, that makes it a lot easier to work with various existing data structures. Nice job!

Also very good to see so many functions from `__init__.py` being separated out into smaller files, that makes it a lot easier to navigate. I did not review the adjacency, file, images and libraries related stuff, assuming that's still the same as before, but just moved.

I only have some smaller comments for a couple of things. One question is perhaps worth discussing more generally: the name of `SequenceDict` instead of `ListDict`. It seems to be a bit inconsistent with the `DictList`, which can in principle also be a sequence. Not sure what your considerations were here? Also curious to hear what others think.

iosonofabio commented on src/igraph/utils.py:
```diff
@@ -35,6 +36,15 @@ def _is_running_in_ipython():
         return False
 
 
+def first(iterable, default=None):
```
`next(iter(edges), 0)` will do

ntamas replied:
I'll remove `first` from the `__all__` dict then.

iosonofabio commented on src/igraph/io/objects.py:
```diff
@@ -0,0 +1,813 @@
+    @param edge_foreign_keys: the name of the attributes in the dicts
+      in the edge data source that contain the source and target
+      vertex names.
+    @param iterative: whether to add the edges to the graph one by one,
```
One could allow for an integer here that acts as a buffer, so you can:

```python
g = Graph.DictList(..., iterative=1000)
```
and it will call `add_edges` every 1000 edges (and once more at the end).

Dealing with all attributes correctly is a little tricky, so I opened a separate issue #436

iosonofabio commented on src/igraph/io/objects.py:
```diff
@@ -0,0 +1,813 @@
+        return result
+
+    # Construct the vertices
+    vertex_attrs, n = {}, 0
```
I've come to like this over time, but many of these are @ntamas babies :smirk: 

I'll just try to avoid double assignments just to keep it simple.

iosonofabio commented on src/igraph/io/objects.py:
```diff
@@ -0,0 +1,813 @@
+    else:
+        vertex_attrs[vertex_name_attr] = []
+
+    vertex_names = vertex_attrs[vertex_name_attr]
```
yep

iosonofabio commented on src/igraph/io/objects.py:
```diff
@@ -0,0 +1,813 @@
+                g.vs[n][vertex_name_attr] = dst_name
+                n += 1
+            g.add_edge(v1, v2)
+            for k, v in edge_data.items():
```
this is all @ntamas 's code but I thought the same. But that's a change of API if we do it now, and it doesn't really hurt. For now, I've clarified the docstring to tell people what to expect.

iosonofabio commented on src/igraph/io/objects.py:
```diff
@@ -0,0 +1,813 @@
+
+        return g
+    else:
+        edge_list, edge_attrs, m = [], {}, 0
```
haha, not so clearly since I haven't written this one :smirk:

In any case, changed now

iosonofabio commented on src/igraph/io/objects.py:
```diff
@@ -0,0 +1,813 @@
+      with the extra items in the edge list (starting from index 2, since
+      the first two items are the source and target vertices). C{None}
+      means that only the source and target vertices will be extracted
+      from each item. If you pass a string here, it will be wrapped in
```
changed

iosonofabio commented on src/igraph/io/objects.py:
```diff
@@ -0,0 +1,813 @@
+      contain the vertex names.
+    @param edge_attrs: the names of the edge attributes that are filled
+      with the extra items in the edge list (starting from index 2, since
+      the first two items are the source and target vertices). C{None}
```
Tried to clarify a bit

iosonofabio commented on src/igraph/io/objects.py:
```diff
@@ -0,0 +1,813 @@
+      means that only the source and target vertices will be extracted
+      from each item. If you pass a string here, it will be wrapped in
+      a list for convenience.
+    @param weights: alternative way to specify that the graph is
```
@ntamas I'll deflect this one to you, as I'm not sure what the original intention here was.

I *suspect* it could be a common situation to have two separate data structures for edge source-targets and for weights, and this argument lifts the need to `zip` them together, which is slow in Python. If that's the case, I think the original rationale was fine but now that we have pandas DataFrame support and a bunch of other input constructors we could get rid of this (in a separate PR, since it requires API deprecation).

If you think that's alright we can open an issue about deprecating this and I'll make a separate PR.

ntamas replied:
`zip` is probably not that slow that it would matter; anyhow, the idea was indeed to make it easy to deal with the common scenario of having an edge list + weights only. I'll leave it as is for the time being; we can go through a deprecation cycle if it really bothers someone but I'm okay with keeping the old API.

iosonofabio commented on src/igraph/io/objects.py:
```diff
@@ -0,0 +1,813 @@
+        for source, sequence in edges.items():
+            n = max(n, source, *sequence)
+            for target in sequence:
+                edge_list.append((source, target))
```
I'm not sure whether or how much more efficient it would be, actually... Python is not C hehe

ntamas replied:
The solution from `@iosonofabio` pushes the entire `max` loop down to the C layer so it's probably faster. One thing that bothers me slightly is that I don't know whether Python has a limit on the maximum number of arguments that you can spread with the `*` operator -- but it seems to work fine with 1M arguments:

```
>>> x = list(range(1000000))
>>> max(-42, *x)
999999
```

ntamas replied:
Actually, I'm going to rewrite the for loop to push down the loop itself to the C layer again; my experience is that an implicit for loop in the C layer (or more) is almost always better than an explicit for loop in the Python layer:

```
edge_list.extend((u, v) for u, v in zip(cycle(source), target))
```

iosonofabio commented on src/igraph/io/objects.py:
```diff
@@ -0,0 +1,813 @@
+
+    for vertex in graph.vs:
+        attrdic = vertex.attributes()
+        if skip_None:
```
hmm, not super sure what you mean but I now put an if/else

iosonofabio commented on src/igraph/io/objects.py:
```diff
@@ -0,0 +1,813 @@
+        if not use_vids:
+            source, target = vs_names[source], vs_names[target]
+        attrdic = edge.attributes()
+        if skip_None:
```
as above

ntamas replied:
Some linters complain about the casing of `skip_None` so I'm just going to use `skip_none`; arguments are `underscore_cased` and this almost always implies all-lowercase.

iosonofabio commented:
Thanks Vincent, I've implemented most suggestions except API breaking changes. For those, I'd say if @ntamas is fine merging this PR then I can make separate ones breaking the API, to keep things tidy.

ntamas commented:
Thanks @iosonofabio ! Can you also update the `CHANGELOG.md` file in the root of the repo? It's okay if you just commit the changes to the `develop` branch.
